File: ts/src/messageBuilders.ts
import { getEventHash } from "nostr-tools"

import {
  CHAT_MESSAGE_KIND,
  EXPIRATION_TAG,
  REACTION_KIND,
  RECEIPT_KIND,
  TYPING_KIND,
  type ExpirationOptions,
  type ReceiptType,
  type Rumor,
} from "./types"
import { resolveExpirationSeconds, upsertExpirationTag } from "./utils"

export const DUMMY_INNER_PUBKEY =
  "0000000000000000000000000000000000000000000000000000000000000000"

export interface RumorBuildOptions {
  pubkey?: string
  createdAt?: number
  nowMs?: number
  tags?: string[][]
  expiration?: ExpirationOptions | null
  ensureMsTag?: boolean
}

export interface RumorBuildInput extends RumorBuildOptions {
  kind: number
  content?: string
}

export function expirationTagForOptions(
  options: ExpirationOptions,
  nowSeconds: number,
): string[] | undefined {
  const expiresAt = resolveExpirationSeconds(options, nowSeconds)
  if (expiresAt === undefined) return undefined
  return [EXPIRATION_TAG, String(expiresAt)]
}

export function appendExpirationTag(
  tags: string[][],
  options: ExpirationOptions,
  nowSeconds: number,
): void {
  const tag = expirationTagForOptions(options, nowSeconds)
  if (tag) upsertExpirationTag(tags, Number(tag[1]))
}

export function ensureMsTag(tags: string[][], ms: number): string[][] {
  const next = cloneTags(tags)
  if (!next.some(([key]) => key === "ms")) {
    next.push(["ms", String(ms)])
  }
  return next
}

export function ensureRecipientTag(tags: string[][], recipientPubkey: string): string[][] {
  const next = cloneTags(tags)
  const hasRecipient = next.some(
    ([key, value]) => key === "p" && value === recipientPubkey,
  )
  if (!hasRecipient) {
    next.unshift(["p", recipientPubkey])
  }
  return next
}

export function buildRumorEvent(input: RumorBuildInput): Rumor {
  const nowMs = input.nowMs ?? Date.now()
  const createdAt = input.createdAt ?? Math.floor(nowMs / 1000)
  let tags = cloneTags(input.tags ?? [])

  if (input.ensureMsTag ?? true) {
    tags = ensureMsTag(tags, nowMs)
  }

  if (input.expiration !== undefined && input.expiration !== null) {
    appendExpirationTag(tags, input.expiration, createdAt)
  }

  const rumor: Rumor = {
    id: "",
    pubkey: input.pubkey ?? DUMMY_INNER_PUBKEY,
    created_at: createdAt,
    kind: input.kind,
    tags,
    content: input.content ?? "",
  }
  rumor.id = getEventHash(rumor)
  return rumor
}

export function buildDirectRumorEvent(
  authorPubkey: string,
  recipientPubkey: string,
  kind: number,
  content: string,
  tags: string[][] = [],
  options: Omit<RumorBuildOptions, "pubkey" | "tags"> = {},
): Rumor {
  return buildRumorEvent({
    ...options,
    pubkey: authorPubkey,
    kind,
    content,
    tags: ensureRecipientTag(tags, recipientPubkey),
  })
}

export function buildTextRumor(
  text: string,
  options: RumorBuildOptions = {},
): Rumor {
  return buildRumorEvent({
    ...options,
    kind: CHAT_MESSAGE_KIND,
    content: text,
  })
}

export function buildReplyRumor(
  text: string,
  replyTo: string,
  options: RumorBuildOptions = {},
): Rumor {
  return buildTextRumor(text, {
    ...options,
    tags: [...cloneTags(options.tags ?? []), eventReferenceTag(replyTo)],
  })
}

export function buildReactionRumor(
  messageId: string,
  emoji: string,
  options: RumorBuildOptions = {},
): Rumor {
  return buildRumorEvent({
    ...options,
    kind: REACTION_KIND,
    content: emoji,
    tags: [...cloneTags(options.tags ?? []), eventReferenceTag(messageId)],
  })
}

export function buildReceiptRumor(
  receiptType: ReceiptType,
  messageIds: string[],
  options: RumorBuildOptions = {},
): Rumor {
  return buildRumorEvent({
    ...options,
    kind: RECEIPT_KIND,
    content: receiptType,
    tags: [
      ...cloneTags(options.tags ?? []),
      ...messageIds.map((id) => eventReferenceTag(id)),
    ],
  })
}

export function buildTypingRumor(options: RumorBuildOptions = {}): Rumor {
  return buildRumorEvent({
    ...options,
    kind: TYPING_KIND,
    content: "typing",
  })
}

export function eventReferenceTag(messageId: string): string[] {
  return ["e", messageId]
}

function cloneTags(tags: string[][]): string[][] {
  return tags.map((tag) => [...tag])
}

File: rust/crates/nostr-double-ratchet/src/message_builders.rs
use std::time::{SystemTime, UNIX_EPOCH};

use nostr::{EventBuilder, Kind, PublicKey, Tag, Timestamp, UnsignedEvent};

use crate::{Error, Result};

#[derive(Debug, Clone, Default)]
pub struct InnerEventBuildOptions {
    pub created_at_seconds: Option<u64>,
    pub ms: Option<u64>,
    pub ensure_ms_tag: bool,
}

impl InnerEventBuildOptions {
    pub fn with_ms_tag() -> Self {
        Self {
            ensure_ms_tag: true,
            ..Self::default()
        }
    }
}

pub fn expiration_tag_for_options(
    options: &crate::SendOptions,
    now_seconds: u64,
) -> Result<Option<Tag>> {
    let Some(expires_at) = crate::utils::resolve_expiration_seconds(options, now_seconds)? else {
        return Ok(None);
    };

    Tag::parse(&[crate::EXPIRATION_TAG.to_string(), expires_at.to_string()])
        .map(Some)
        .map_err(|e| Error::InvalidEvent(e.to_string()))
}

pub fn append_expiration_tag(
    tags: &mut Vec<Tag>,
    options: &crate::SendOptions,
    now_seconds: u64,
) -> Result<()> {
    if let Some(tag) = expiration_tag_for_options(options, now_seconds)? {
        tags.push(tag);
    }
    Ok(())
}

pub fn build_inner_event(
    author_pubkey: PublicKey,
    kind: u32,
    content: impl Into<String>,
    tags: Vec<Tag>,
    options: InnerEventBuildOptions,
) -> Result<UnsignedEvent> {
    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
    let now_seconds = now.as_secs();
    let now_ms = now.as_millis() as u64;
    let created_at_seconds = options
        .created_at_seconds
        .or_else(|| options.ms.map(|ms| ms / 1_000))
        .unwrap_or(now_seconds);
    let ms = options.ms.unwrap_or(now_ms);
    let tags = if options.ensure_ms_tag {
        ensure_ms_tag(tags, ms)?
    } else {
        tags
    };

    let kind: u16 = kind
        .try_into()
        .map_err(|_| Error::InvalidEvent("kind out of range".to_string()))?;
    let mut event = EventBuilder::new(Kind::from(kind), content.into())
        .tags(tags)
        .custom_created_at(Timestamp::from(created_at_seconds))
        .build(author_pubkey);
    event.ensure_id();
    Ok(event)
}

pub fn build_direct_inner_event(
    author_pubkey: PublicKey,
    recipient_pubkey: PublicKey,
    kind: u32,
    content: impl Into<String>,
    tags: Vec<Tag>,
) -> Result<UnsignedEvent> {
    build_inner_event(
        author_pubkey,
        kind,
        content,
        ensure_recipient_tag(tags, recipient_pubkey)?,
        InnerEventBuildOptions::with_ms_tag(),
    )
}

pub fn build_text_rumor(
    author_pubkey: PublicKey,
    text: impl Into<String>,
    tags: Vec<Tag>,
) -> Result<UnsignedEvent> {
    build_inner_event(
        author_pubkey,
        crate::CHAT_MESSAGE_KIND,
        text,
        tags,
        InnerEventBuildOptions::with_ms_tag(),
    )
}

pub fn build_reply_rumor(
    author_pubkey: PublicKey,
    text: impl Into<String>,
    reply_to: &str,
    mut tags: Vec<Tag>,
) -> Result<UnsignedEvent> {
    tags.push(event_reference_tag(reply_to)?);
    build_text_rumor(author_pubkey, text, tags)
}

pub fn build_reaction_rumor(
    author_pubkey: PublicKey,
    message_id: &str,
    emoji: impl Into<String>,
    mut tags: Vec<Tag>,
) -> Result<UnsignedEvent> {
    tags.push(event_reference_tag(message_id)?);
    build_inner_event(
        author_pubkey,
        crate::REACTION_KIND,
        emoji,
        tags,
        InnerEventBuildOptions::with_ms_tag(),
    )
}

pub fn build_receipt_rumor(
    author_pubkey: PublicKey,
    receipt_type: impl Into<String>,
    message_ids: impl IntoIterator<Item = impl AsRef<str>>,
    mut tags: Vec<Tag>,
) -> Result<UnsignedEvent> {
    for message_id in message_ids {
        tags.push(event_reference_tag(message_id.as_ref())?);
    }
    build_inner_event(
        author_pubkey,
        crate::RECEIPT_KIND,
        receipt_type,
        tags,
        InnerEventBuildOptions::with_ms_tag(),
    )
}

pub fn build_typing_rumor(author_pubkey: PublicKey, tags: Vec<Tag>) -> Result<UnsignedEvent> {
    build_inner_event(
        author_pubkey,
        crate::TYPING_KIND,
        "typing",
        tags,
        InnerEventBuildOptions::with_ms_tag(),
    )
}

pub fn ensure_recipient_tag(mut tags: Vec<Tag>, recipient_pubkey: PublicKey) -> Result<Vec<Tag>> {
    let recipient_hex = recipient_pubkey.to_hex();
    let has_recipient_p_tag = tags.iter().any(|tag| {
        let values = tag.clone().to_vec();
        values.first().map(|value| value.as_str()) == Some("p")
            && values.get(1).map(|value| value.as_str()) == Some(recipient_hex.as_str())
    });

    if !has_recipient_p_tag {
        tags.insert(
            0,
            Tag::parse(&["p".to_string(), recipient_hex])
                .map_err(|e| Error::InvalidEvent(e.to_string()))?,
        );
    }

    Ok(tags)
}

pub fn ensure_ms_tag(mut tags: Vec<Tag>, ms: u64) -> Result<Vec<Tag>> {
    let has_ms_tag = tags.iter().any(|tag| {
        let values = tag.clone().to_vec();
        values.first().map(|value| value.as_str()) == Some("ms")
    });

    if !has_ms_tag {
        tags.push(
            Tag::parse(&["ms".to_string(), ms.to_string()])
                .map_err(|e| Error::InvalidEvent(e.to_string()))?,
        );
    }

    Ok(tags)
}

pub fn event_reference_tag(message_id: &str) -> Result<Tag> {
    Tag::parse(&["e".to_string(), message_id.to_string()])
        .map_err(|e| Error::InvalidEvent(e.to_string()))
}
